common/scpi.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <compiler.h>
+#include <debug.h>
+#include <dm.h>
+#include <error.h>
+#include <limits.h>
+#include <scpi.h>
+#include <scpi_internal.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <string.h>
+#include <work.h>
+#include <drivers/msgbox.h>
+#include <drivers/wallclock.h>
+#include <platform/memory.h>
+
+#define SCPI_MEM_AREA(n) (((struct scpi_mem *)SCPI_MEM_BASE)[n])
+
+/* 100 ms */
+#define SCPI_TIMEOUT     100 * 1000 * 24
+
+struct scpi_mem {
+	struct scpi_msg tx_msg;
+	struct scpi_msg rx_msg;
+} __packed;
+
+static struct scpi_mem scpi_buffer;
+static struct device  *scpi_msgbox;
+
+static bool scpi_buffer_busy;
+
+static inline void *
+chan_to_param(uint8_t chan)
+{
+	return (void *)(uintptr_t)chan;
+}
+
+static inline uint8_t
+param_to_chan(void *param)
+{
+	return (uint8_t)(uintptr_t)param;
+}
+
+static void
+scpi_send_message(void *param)
+{
+	int      err;
+	uint8_t  chan    = param_to_chan(param);
+	uint64_t timeout = wallclock_read() + SCPI_TIMEOUT;
+
+	/* Wait for previous message to be acknowledged, with a timeout. */
+	while (msgbox_tx_pending(scpi_msgbox, chan)) {
+		if (wallclock_read() >= timeout) {
+			warn("SCPI:%u: Channel busy, retrying send", chan);
```

Comment:
From an API standpoint, wouldn't it make more sense to return `EAGAIN` rather than trying to resend the message on timeout?

Reply:
No, this is internal, and it's a work function, so it can't return anything. I can add more comments explaining this, but the reason is that we want to try as hard as possible to avoid dropping messages.

Reply:
That makes sense; thanks for the explanation.